### ISSUE TITLE
multiple definition error with gcc v10.2.0

### ISSUE THREAD

**Issue by adahbingee** [FreeGLUTProject/freeglut#86]
**Windows 10 with msys2**

```
[ 31%] Linking C shared library bin\libfreeglut.dll
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_gl2.c.obj):fg_gl2.c:(.bss+0x0): multiple definition of `fghGenBuffers'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x0): first defined here
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_gl2.c.obj):fg_gl2.c:(.bss+0x8): multiple definition of `fghDeleteBuffers'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x8): first defined here
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_gl2.c.obj):fg_gl2.c:(.bss+0x10): multiple definition of `fghBindBuffer'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x10): first defined here
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_gl2.c.obj):fg_gl2.c:(.bss+0x18): multiple definition of `fghBufferData'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x18): first defined here
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_gl2.c.obj):fg_gl2.c:(.bss+0x20): multiple definition of `fghEnableVertexAttribArray'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x20): first defined here
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_gl2.c.obj):fg_gl2.c:(.bss+0x28): multiple definition of `fghDisableVertexAttribArray'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x28): first defined here
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_gl2.c.obj):fg_gl2.c:(.bss+0x30): multiple definition of `fghVertexAttribPointer'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x30): first defined here
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_window.c.obj):fg_window.c:(.bss+0x0): multiple definition of `fghGenBuffers'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x0): first defined here
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_window.c.obj):fg_window.c:(.bss+0x8): multiple definition of `fghDeleteBuffers'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x8): first defined here
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_window.c.obj):fg_window.c:(.bss+0x10): multiple definition of `fghBindBuffer'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x10): first defined here
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_window.c.obj):fg_window.c:(.bss+0x18): multiple definition of `fghBufferData'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x18): first defined here
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_window.c.obj):fg_window.c:(.bss+0x20): multiple definition of `fghEnableVertexAttribArray'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x20): first defined here
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_window.c.obj):fg_window.c:(.bss+0x28): multiple definition of `fghDisableVertexAttribArray'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x28): first defined here
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_window.c.obj):fg_window.c:(.bss+0x30): multiple definition of `fghVertexAttribPointer'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x30): first defined here
collect2.exe: error: ld returned 1 exit status
make[2]: *** [CMakeFiles\freeglut.dir\build.make:704: bin/libfreeglut.dll] Error 1
make[1]: *** [CMakeFiles\Makefile2:171: CMakeFiles/freeglut.dir/all] Error 2
make: *** [Makefile:148: all] Error 2
```

**mingw64 version**
```
$ gcc -v
Using built-in specs.
COLLECT_GCC=C:\mingw64\bin\gcc.exe
COLLECT_LTO_WRAPPER=C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: ../gcc-10.2.0/configure --prefix=/mingw64 --with-local-prefix=/mingw64/local --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --with-native-system-header-dir=/mingw64/x86_64-w64-mingw32/include --libexecdir=/mingw64/lib --enable-bootstrap --with-arch=x86-64 --with-tune=generic --enable-languages=c,lto,c++,fortran,ada,objc,obj-c++ --enable-shared --enable-static --enable-libatomic --enable-threads=posix --enable-graphite --enable-fully-dynamic-string --enable-libstdcxx-filesystem-ts=yes --enable-libstdcxx-time=yes --disable-libstdcxx-pch --disable-libstdcxx-debug --disable-isl-version-check --enable-lto --enable-libgomp --disable-multilib --enable-checking=release --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --disable-plugin --with-libiconv --with-system-zlib --with-gmp=/mingw64 --with-mpfr=/mingw64 --with-mpc=/mingw64 --with-isl=/mingw64 --with-pkgversion='Rev1, Built by MSYS2 project' --with-bugurl=https://github.com/msys2/MINGW-packages/issues --with-gnu-as --with-gnu-ld
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 10.2.0 (Rev1, Built by MSYS2 project)

```

**cmake**
```
The C compiler identification is GNU 10.2.0
The CXX compiler identification is GNU 10.2.0
Detecting C compiler ABI info
Detecting C compiler ABI info - done
Check for working C compiler: C:/mingw64/bin/gcc.exe - skipped
Detecting C compile features
Detecting C compile features - done
Detecting CXX compiler ABI info
Detecting CXX compiler ABI info - done
Check for working CXX compiler: C:/mingw64/bin/g++.exe - skipped
Detecting CXX compile features
Detecting CXX compile features - done
Found OpenGL: opengl32   
Looking for include file sys/types.h
Looking for include file sys/types.h - found
Looking for include file unistd.h
Looking for include file unistd.h - found
Looking for include file sys/time.h
Looking for include file sys/time.h - found
Looking for include file stdbool.h
Looking for include file stdbool.h - found
Looking for include file sys/param.h
Looking for include file sys/param.h - found
Looking for include file sys/ioctl.h
Looking for include file sys/ioctl.h - not found
Looking for include file fcntl.h
Looking for include file fcntl.h - found
Looking for include file usbhid.h
Looking for include file usbhid.h - not found
Looking for gettimeofday
Looking for gettimeofday - found
Looking for XParseGeometry
Looking for XParseGeometry - not found
Looking for include file stdint.h
Looking for include file stdint.h - found
Looking for include file inttypes.h
Looking for include file inttypes.h - found
Configuring done
Generating done
```

**Comment by Arin112** [comment]
The issue can be fixed with adding `-fcommon` to compiler options.

> A common mistake in C is omitting extern when declaring a global variable in a header file. If the header is included by several files it results in multiple definitions of the same variable. In previous GCC versions this error is ignored. GCC 10 defaults to -fno-common, which means a linker error will now be reported. To fix this, use extern in header files when declaring global variables, and ensure each global is defined in exactly one C file. If tentative definitions of particular variables need to be placed in a common block, __attribute__((__common__)) can be used to force that behavior even in code compiled without -fcommon. As a workaround, legacy C code where all tentative definitions should be placed into a common block can be compiled with -fcommon.
[https://gcc.gnu.org/gcc-10/porting_to.html](url)

**Comment by adahbingee** [comment]
Why not fix this common mistake?

**Comment by tomatenkuchen** [comment]
Hi!
I tested freeglut 3.11. It's the latest in the github repo and tried compilation with gcc12 and clang14. Both run like a charm. No linker issues. Seems that they fixed it in the latest version. If conan-center provides the new version, you can omit the verification step with the compiler version check.

**Comment by tomwillow** [comment]
> **Windows 10 with msys2**
> 
> ```
> [ 31%] Linking C shared library bin\libfreeglut.dll
> C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_gl2.c.obj):fg_gl2.c:(.bss+0x0): multiple definition of `fghGenBuffers'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x0): first defined here
> C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_gl2.c.obj):fg_gl2.c:(.bss+0x8): multiple definition of `fghDeleteBuffers'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x8): first defined here
> C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_gl2.c.obj):fg_gl2.c:(.bss+0x10): multiple definition of `fghBindBuffer'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x10): first defined here
> C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_gl2.c.obj):fg_gl2.c:(.bss+0x18): multiple definition of `fghBufferData'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x18): first defined here
> C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_gl2.c.obj):fg_gl2.c:(.bss+0x20): multiple definition of `fghEnableVertexAttribArray'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x20): first defined here
> C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_gl2.c.obj):fg_gl2.c:(.bss+0x28): multiple definition of `fghDisableVertexAttribArray'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x28): first defined here
> C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_gl2.c.obj):fg_gl2.c:(.bss+0x30): multiple definition of `fghVertexAttribPointer'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x30): first defined here
> C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_window.c.obj):fg_window.c:(.bss+0x0): multiple definition of `fghGenBuffers'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x0): first defined here
> C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_window.c.obj):fg_window.c:(.bss+0x8): multiple definition of `fghDeleteBuffers'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x8): first defined here
> C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_window.c.obj):fg_window.c:(.bss+0x10): multiple definition of `fghBindBuffer'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x10): first defined here
> C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_window.c.obj):fg_window.c:(.bss+0x18): multiple definition of `fghBufferData'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x18): first defined here
> C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_window.c.obj):fg_window.c:(.bss+0x20): multiple definition of `fghEnableVertexAttribArray'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x20): first defined here
> C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_window.c.obj):fg_window.c:(.bss+0x28): multiple definition of `fghDisableVertexAttribArray'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x28): first defined here
> C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\freeglut.dir/objects.a(fg_window.c.obj):fg_window.c:(.bss+0x30): multiple definition of `fghVertexAttribPointer'; CMakeFiles\freeglut.dir/objects.a(fg_geometry.c.obj):fg_geometry.c:(.bss+0x30): first defined here
> collect2.exe: error: ld returned 1 exit status
> make[2]: *** [CMakeFiles\freeglut.dir\build.make:704: bin/libfreeglut.dll] Error 1
> make[1]: *** [CMakeFiles\Makefile2:171: CMakeFiles/freeglut.dir/all] Error 2
> make: *** [Makefile:148: all] Error 2
> ```
> 
> **mingw64 version**
> 
> ```
> $ gcc -v
> Using built-in specs.
> COLLECT_GCC=C:\mingw64\bin\gcc.exe
> COLLECT_LTO_WRAPPER=C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/lto-wrapper.exe
> Target: x86_64-w64-mingw32
> Configured with: ../gcc-10.2.0/configure --prefix=/mingw64 --with-local-prefix=/mingw64/local --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --with-native-system-header-dir=/mingw64/x86_64-w64-mingw32/include --libexecdir=/mingw64/lib --enable-bootstrap --with-arch=x86-64 --with-tune=generic --enable-languages=c,lto,c++,fortran,ada,objc,obj-c++ --enable-shared --enable-static --enable-libatomic --enable-threads=posix --enable-graphite --enable-fully-dynamic-string --enable-libstdcxx-filesystem-ts=yes --enable-libstdcxx-time=yes --disable-libstdcxx-pch --disable-libstdcxx-debug --disable-isl-version-check --enable-lto --enable-libgomp --disable-multilib --enable-checking=release --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --disable-plugin --with-libiconv --with-system-zlib --with-gmp=/mingw64 --with-mpfr=/mingw64 --with-mpc=/mingw64 --with-isl=/mingw64 --with-pkgversion='Rev1, Built by MSYS2 project' --with-bugurl=https://github.com/msys2/MINGW-packages/issues --with-gnu-as --with-gnu-ld
> Thread model: posix
> Supported LTO compression algorithms: zlib zstd
> gcc version 10.2.0 (Rev1, Built by MSYS2 project)
> ```
> 
> **cmake**
> 
> ```
> The C compiler identification is GNU 10.2.0
> The CXX compiler identification is GNU 10.2.0
> Detecting C compiler ABI info
> Detecting C compiler ABI info - done
> Check for working C compiler: C:/mingw64/bin/gcc.exe - skipped
> Detecting C compile features
> Detecting C compile features - done
> Detecting CXX compiler ABI info
> Detecting CXX compiler ABI info - done
> Check for working CXX compiler: C:/mingw64/bin/g++.exe - skipped
> Detecting CXX compile features
> Detecting CXX compile features - done
> Found OpenGL: opengl32   
> Looking for include file sys/types.h
> Looking for include file sys/types.h - found
> Looking for include file unistd.h
> Looking for include file unistd.h - found
> Looking for include file sys/time.h
> Looking for include file sys/time.h - found
> Looking for include file stdbool.h
> Looking for include file stdbool.h - found
> Looking for include file sys/param.h
> Looking for include file sys/param.h - found
> Looking for include file sys/ioctl.h
> Looking for include file sys/ioctl.h - not found
> Looking for include file fcntl.h
> Looking for include file fcntl.h - found
> Looking for include file usbhid.h
> Looking for include file usbhid.h - not found
> Looking for gettimeofday
> Looking for gettimeofday - found
> Looking for XParseGeometry
> Looking for XParseGeometry - not found
> Looking for include file stdint.h
> Looking for include file stdint.h - found
> Looking for include file inttypes.h
> Looking for include file inttypes.h - found
> Configuring done
> Generating done
> ```

My environment is like as you. My freeglut version is 3.2.1, and when I use Mingw64 in msys2 win10x64 to link my code, the same error messages is printed.

I found a method for solving it:
1. After cmake gui.exe "configure" "generate" buttons pressing, enter this path:
"D:\...\freeglut-3.2.1\x64_makefile_out\CMakeFiles\freeglut_static.dir" where "x64_makefile_out" is your output directory.
2. Open the flags.make file.
modify its 6th line: "C_FLAGS =  -Wall -pedantic" tobe "C_FLAGS =  -Wall -pedantic  -fcommon".
3. Then build as normal.
$ cd x64_makefile_out
$ cmake --build .
4. Now all is well.

**Comment by jtsiomb** [comment]
This is not the official freeglut repo. It's a mirror. The fix for this has been in the freeglut SVN repo for a long time now, and it's also included in the latest release: 3.2.2. This issue should be closed.
